### PR TITLE
Refactor form verifiers API.

### LINKED ISSUE #9696: amp-form: async verification - design changes
Issue to track incorporating design changes as discussed among @aghassemi @cvializ @dvoytenko

### PR DESCRIPTION
**Pull request by cvializ**
This implements the suggestions @dvoytenko gave for the form verification API.

- No config block, no verification groups
- Verification behavior is triggered by a `verify-xhr` attribute functioning similarly to `action-xhr`
- The verification request must be sent after every `change` event. The groups allowed us to limit the requests to only send when a group was complete

Closes #9696

### CODE REVIEW

File: extensions/amp-form/0.1/amp-form.js [hunk 1]
```diff
@@ -242,27 +247,32 @@ export class AmpForm {
   installEventHandlers_() {
     this.form_.addEventListener(
         'submit', this.handleSubmitEvent_.bind(this), true);
+
     this.form_.addEventListener('blur', e => {
       checkUserValidityAfterInteraction_(dev().assertElement(e.target));
       this.validator_.onBlur(e);
     }, true);
+
+    const cleanup = () => {
```
Comment (aghassemi) [inline extensions/amp-form/0.1/amp-form.js]: `cleanup` too generic given this methods deals with non-verifier code as well. Maybe `afterVerifierCommit` ?

File: extensions/amp-form/0.1/amp-form.js [hunk 2]
```diff
@@ -438,28 +450,49 @@ export class AmpForm {
...
-    const isHeadOrGet = this.method_ == 'GET' || this.method_ == 'HEAD';
-    const values = this.getFormAsObject_(opt_extraValues);
-    this.renderTemplate_(values);
+  doActionXhr_() {
+    return this.doXhr_(dev().assertString(this.xhrAction_), this.method_);
+  }
 
+  /**
+   * Send a request to the form's verify endpoint.
+   * @private
```
Comment (aghassemi) [inline extensions/amp-form/0.1/amp-form.js]: add return type.

File: extensions/amp-form/0.1/amp-form.js [hunk 3]
```diff
@@ -438,28 +450,49 @@ export class AmpForm {
...
+    return this.doXhr_(dev().assertString(this.xhrVerify_), this.method_,
+        {[FORM_VERIFY_PARAM]: true});
+  }
+
+  /**
+   * Send a request to a form endpoint.
+   * @param {string} url
+   * @param {string} method
+   * @param {!Object<string, string>=} opt_extraFields
+   * @private
```
Comment (aghassemi) [inline extensions/amp-form/0.1/amp-form.js]: ditto return type

File: extensions/amp-form/0.1/form-verifiers.js [hunk 4]
```diff
@@ -133,30 +62,82 @@ function isSubmittable_(element) {
...
-   * e.g. the input's 'change' event
-   * @param {!Element} unusedInput
-   * @param {!function(!Array<!Element>)} unusedAfterVerify
+   * Checks if the form has been changed from its initial state.
+   * @private
+   */
+  isDirty_() {
+    const form = this.form_;
+    for (let i = 0; i < form.elements.length; i++) {
+      const field = form.elements[i];
```
Comment (aghassemi) [inline extensions/amp-form/0.1/form-verifiers.js]: I think we can break early for disable elements, they won't get submitted anyway.

Comment (cvializ) [inline extensions/amp-form/0.1/form-verifiers.js]: Done.

File: extensions/amp-form/0.1/form-verifiers.js [hunk 5]
```diff
@@ -133,30 +62,85 @@ function isSubmittable_(element) {
...
```
Comment (jridgewell) [inline extensions/amp-form/0.1/form-verifiers.js]: Whoo, this is going to be a slow loop. This'll recreate `elements` HTMLFormControlsCollection on every iteration.

Comment (cvializ) [inline extensions/amp-form/0.1/form-verifiers.js]: Ah I didn't consider it'd be computed every access. I'll cache it in a variable outside the loop.

Comment (cvializ) [inline extensions/amp-form/0.1/form-verifiers.js]: #9884

**Pull request merged**